src/dialects/abstract/data-types.ts

```diff
@@ -0,0 +1,1826 @@
+      throw new ValidationError(
+        util.format('%j is not a valid range', value),
+        [],
+      );
+    }
+
+    if (value.length !== 2) {
+      throw new ValidationError(
+        'A range must be an array with two elements',
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * Range types are data types representing a range of values of some element type (called the range's subtype).
+ * Only available in Postgres. See [the Postgres documentation](http://www.postgresql.org/docs/9.4/static/rangetypes.html) for more details
+ */
+export type RANGE<T extends NUMBER | DATE | DATEONLY> = _RANGE<T>;
+export const RANGE = classToInvokable(_RANGE);
+
+interface UUIDOptions {
+  acceptStrings?: boolean;
+}
+
+class _UUID extends ABSTRACT<string> {
+  public static readonly key = 'UUID';
+
+  public validate(value: string, options?: UUIDOptions) {
+    if (typeof value !== 'string') {
+      throw new ValidationError(
+        util.format('%j is not a valid uuid', value),
+        [],
+      );
+    }
+
+    if (!options?.acceptStrings && !Validator.isUUID(value)) {
+      throw new ValidationError(
+        util.format('%j is not a valid uuid', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * A column storing a unique universal identifier.
+ * Use with `UUIDV1` or `UUIDV4` for default values.
+ */
+export type UUID = _UUID;
+export const UUID = classToInvokable(_UUID);
+
+class _UUIDV1 extends ABSTRACT<string> {
+  public static readonly key = 'UUIDV1';
+
+  public validate(value: string, options?: UUIDOptions) {
+    if (typeof value !== 'string') {
+      throw new ValidationError(
+        util.format('%j is not a valid uuid', value),
+        [],
+      );
+    }
+
+    if (!options?.acceptStrings && !Validator.isUUID(value, 1)) {
+      throw new ValidationError(
+        util.format('%j is not a valid uuidv1', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * A default unique universal identifier generated following the UUID v1 standard
+ */
+export type UUIDV1 = _UUIDV1;
+export const UUIDV1 = classToInvokable(_UUIDV1);
+
+class _UUIDV4 extends ABSTRACT<string> {
+  public static readonly key = 'UUIDV4';
+
+  public validate(value: string, options?: UUIDOptions) {
+    if (typeof value !== 'string') {
+      throw new ValidationError(
+        util.format('%j is not a valid uuid', value),
+        [],
+      );
+    }
+
+    if (!options?.acceptStrings && !Validator.isUUID(value, 4)) {
+      throw new ValidationError(
+        util.format('%j is not a valid uuidv4', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * A default unique universal identifier generated following the UUID v4 standard
+ */
+export type UUIDV4 = _UUIDV4;
+export const UUIDV4 = classToInvokable(_UUIDV4);
+
+class _VIRTUAL<T> extends ABSTRACT<T> {
+  public static readonly key = 'VIRTUAL';
+
+  returnType?: ABSTRACT<T>;
+  fields?: string[];
```

Comment:
```suggestion
  returnType: ABSTRACT<T> | undefined;
  fields: string[] | undefined;
```

these fields are always created, but they can be undefiend